client/landing/gutenboarding/stores/domain-suggestions/utils.ts

```diff
@@ -0,0 +1,19 @@
+/**
+ * External dependencies
+ */
+import deterministicStringify from 'fast-json-stable-stringify';
+
+/**
+ * Internal dependencies
+ */
+import { DomainSuggestionQuery } from './types';
+
+/**
+ * Stable transform to an object key for storage and access.
+ *
+ * @TODO Should we eliminate some properties from serialization like `getSerializedDomainsSuggestionsQuery`?
+ * @see client/state/domains/suggestions/utils.js
+ */
+export const stringifyDomainQueryObject: (
+	q: DomainSuggestionQuery
+) => string = deterministicStringify;
```

Comment:
👋 @Automattic/cobalt Can you help us understand how closely this logic should be respected?

https://github.com/Automattic/wp-calypso/blob/dca7aec0efa149aca53de76dcf7347eb2cf74f93/client/state/domains/suggestions/utils.js#L27-L32

We normalize the query elsewhere:

https://github.com/Automattic/wp-calypso/blob/dca7aec0efa149aca53de76dcf7347eb2cf74f93/client/landing/gutenboarding/stores/domain-suggestions/selectors.ts#L49-L63

But should we omit properties? Are they irrelevant for the domain suggestions endpoint?

Reply:
Not sure I understand - omit which properties? All the parameters we pass to the suggestions endpoint should be relevant and needed.

Reply:
> All the parameters we pass to the suggestions endpoint should be relevant and needed.

That doesn't seem to be the case right now. This query component builds the query sent through the API:

https://github.com/Automattic/wp-calypso/blob/dca7aec0efa149aca53de76dcf7347eb2cf74f93/client/components/data/query-domains-suggestions/index.jsx#L17-L26

But this is how the query is serialized for storage:

https://github.com/Automattic/wp-calypso/blob/dca7aec0efa149aca53de76dcf7347eb2cf74f93/client/state/domains/suggestions/utils.js#L27-L32

Note that this serialization function —which transforms the query object into a string for storage and retrieval from state— it seems to ignore the following properties:

- `recommendation_context`
- `vertical`

That would lead to collisions for queries that only differ on those properties. It sounds like this may be an existing bug which would be corrected with this new data store.

Reply:
Also, anecdotally: I tested e.g. with `vertical` set to one of those funky `p12v2` IDs, but domain suggestions didn't seem to differ regardless of whether I passed it or didn't -- so I think the question is, does the endpoint actually do anything with `vertical`?

About `quantity` and `vendor`: The endpoint works fine without them (and apparently assumes some defaults, e.g. `5` for `quantity`), so it's very surprising to see these fields enforced as required at a totally different part of the app.

Reply:
It is most likely a bug - the conversion to a Redux store was done by a different team. And seems we're doing it again with Gutenberg.

The endpoint does have some default values for missing parameters, but we should still allow the frontend code to overwrite them. I have no idea how this code is going to be used, to be honest, so I can only say that I'd prefer if the list of parameters the endpoint receives stays the same.

Reply:
> And seems we're doing it again with Gutenberg.

Aside, we plan to plug ours back into Calypso to remove duplication :grimacing:

Reply:
> The endpoint does have some default values for missing parameters, but we should still allow the frontend code to overwrite them. I have no idea how this code is going to be used, to be honest, so I can only say that I'd prefer if the list of parameters the endpoint receives stays the same.

@klimeryk The plan is to keep the API calls and the logic pretty much the same while extracting the functionality to an isolated store that can then be use inside and outside of Calypso.

We'd be happy to fill you in with more details and have you involved in the entire process. Perhaps I should have as part of this issue or before.

The core of what's happening in this store can be summed up as follows:

- **This stores public API**: Consumers select from the store (no concept of query component, API call is implicit):
  ```js
  select( 'automattic/domains/suggestions' ).getDomainSuggestions( 'cats' /* <-- suggestions query */ )`
  ```
- **Internals**: Normalize a query for the API
  https://github.com/Automattic/wp-calypso/blob/69dde911636e6f6c8212a06555ce221b0f473832/client/landing/gutenboarding/stores/domain-suggestions/selectors.ts#L61-L72
- **Internals**: Query the API:
  https://github.com/Automattic/wp-calypso/blob/69dde911636e6f6c8212a06555ce221b0f473832/client/landing/gutenboarding/stores/domain-suggestions/resolvers.ts#L16-L24

We've attempted to match the logic around the data request, while smoothing over rough edges for using the selector to fetch data. If you have feedback or would like to chat more about this, happy to do that here or via another channel.

It's on our near-term roadmap to experiment with removing domains suggestions from the Calypso redux store and replacing it with this. That seems feasible at this time, but we haven't made the attempt yet. We'll be sure to keep you in the loop 👍

Reply:
Hahaha, the thing is we have not moved fully from Flux stores (though, I've seen Team Calypso working hard on that so maybe we finally did?), Redux is not used universally in domains, so I'm not excited by the prospect of another data management paradigm. But I understand _why_ it's being done, so, of course, I want to help.

But at this point, to be honest, I'm mostly interested in not introducing regressions - happy to help with testing and fill in the gaps in _domain_ knowledge (😉), but I'll leave the implementation details to you, since you definitely have more experience and knowledge to tackle it. Please do keep including us (Cobalt) in the loop 🙇

Reply:
> Hahaha, the thing is we have not moved fully from Flux stores (though, I've seen Team Calypso working hard on that so maybe we finally did?), Redux is not used universally in domains, so I'm not excited by the prospect of another data management paradigm. 

Generally, I share the sentiment; I'm always skeptical about introducting a new paradigm when even a legacy one hasn't been fully wiped out yet. In this case however, there are some (IMO) compelling reasons to do this; it has long been identified that Redux state can be too monolithic, and a real performance bottleneck, for an app the size of Calypso. @sgomes and @Automattic/team-calypso have been looking into breaking state into more manageable sized pieces for a while.

Another point is that a lot doesn't change: `@wordpress/data` style stores are, after all, wrappers around Redux (but with smaller-scoped stores), i.e. they use the familiar paradigms of reducers, actions, and selectors; it's just the query components that are going away in favor of resolvers (which are arguably a more elegant solution to the same problem).

Finally, I think the state of application state in 2019, some 5 years into React being around, just goes to show that it's a hard problem to solve, to which -- unlike React -- no more-or-less universally recognized solutions exist. `@wordpress/data`, too, might not be the end-all -- a lot has been written about tradeoffs of possible solutions on Calypso P2 -- but I'm afraid our best shot is to keep experimenting to meet the requirements we have for application state management.

The concrete motivation in this particular case is that we want to keep onboarding lightweight, i.e. not pull in the behemoth that is Calypso state. And given the choice between duplicating that logic vs retrofitting Calypso to use the same store, I think the latter will ultimately provide us with better maintainability and insights into potential approaches to this.

> But I understand _why_ it's being done, so, of course, I want to help.

Thanks, appreciate your understanding a lot!

> But at this point, to be honest, I'm mostly interested in not introducing regressions - happy to help with testing and fill in the gaps in _domain_ knowledge (), but I'll leave the implementation details to you, since you definitely have more experience and knowledge to tackle it. Please do keep including us (Cobalt) in the loop

Again, thanks! From a quick look, the domains suggestions Calypso state implementation didn't look very exotic, so I'm cautiously optimistic that we'll be able to pull this off. (I think we've already identified the trickier bits, which are mostly around defaults, serialization/normalization, and the `vendor` value coming from a lib, and I think we have solid ideas on how to tackle most of them, thanks to your feedback and input.)

We'll keep you in the loop!

Reply:
Thank you for the extra insights - they're very valuable 🙇 Fingers crossed for this approach! 🤞